File: java-worker/src/Worker.java
import java.net.*;
import java.io.*;
import java.util.concurrent.*;

public class Worker {
    private int port;
    public String workerId;
    public RaftConsensus raft;
    public Storage storage;
    private HttpMonitor monitor;
    private ExecutorService executor;
    private ServerSocket serverSocket;
    private volatile boolean running = true;
    
    public Worker(String workerId, int port, int monitorPort) {
        this.workerId = workerId;
        this.port = port;
        this.executor = Executors.newCachedThreadPool();
        this.storage = new Storage(workerId);
        this.raft = new RaftConsensus(workerId, port);
        this.monitor = new HttpMonitor(monitorPort, this);
    }
    
    private void configurePeers() {
        if (workerId.equals("java1")) {
            raft.addPeer("java2", "java-worker-2", 5001);
            raft.addPeer("python1", "python-worker-1", 5002);
            raft.addPeer("python2", "python-worker-2", 5003);
        } else if (workerId.equals("java2")) {
            raft.addPeer("java1", "java-worker-1", 5000);
            raft.addPeer("python1", "python-worker-1", 5002);
            raft.addPeer("python2", "python-worker-2", 5003);
        }
    }

    public void start() {
        configurePeers();
        
        // Iniciar monitor HTTP
        executor.submit(() -> {
            try {
                monitor.start();
            } catch (Exception e) {
                System.err.println("Error en monitor HTTP: " + e.getMessage());
            }
        });
        
        // Iniciar módulo RAFT
        executor.submit(() -> {
            try {
                raft.start();
            } catch (Exception e) {
                System.err.println("Error en RAFT: " + e.getMessage());
            }
        });
        
        // Esperar un poco para que RAFT se estabilice
        try {
            Thread.sleep(3000);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
        
        // Iniciar servidor principal con reintentos
        int retries = 0;
        while (retries < 5 && running) {
            try {
                serverSocket = new ServerSocket();
                serverSocket.setReuseAddress(true);
                serverSocket.bind(new InetSocketAddress("0.0.0.0", port), 50);
                System.out.println("Worker " + workerId + " iniciado exitosamente en puerto " + port);
                break;
            } catch (IOException e) {
                retries++;
                System.err.println("Intento " + retries + " fallido: " + e.getMessage());
                if (retries < 5) {
                    try {
                        Thread.sleep(2000);
                    } catch (InterruptedException ie) {
                        Thread.currentThread().interrupt();
                        return;
                    }
                } else {
                    System.err.println("No se pudo iniciar el servidor después de 5 intentos");
                    return;
                }
            }
        }
        
        // Loop principal del servidor
        while (running) {
            try {
                Socket client = serverSocket.accept();
                executor.submit(() -> handleClient(client));
            } catch (IOException e) {
                if (running) {
                    System.err.println("Error aceptando conexión: " + e.getMessage());
                }
            }
        }
    }
    
    private void handleClient(Socket client) {
        try {
            client.setSoTimeout(10000); // 10 segundos timeout
            
            BufferedReader in = new BufferedReader(
                new InputStreamReader(client.getInputStream(), "UTF-8")
            );
            PrintWriter out = new PrintWriter(
                new OutputStreamWriter(client.getOutputStream(), "UTF-8"), true
            );
            
            String request = in.readLine();
            
            if (request == null) {
                out.println("ERROR|NULL_REQUEST");
                return;
            }
            
            request = request.trim();
            
            if (request.isEmpty()) {
                out.println("ERROR|EMPTY_REQUEST");
                return;
            }
            
            System.out.println("\n=== Nueva petición en " + workerId + " ===");
            System.out.println("Request completo: '" + request + "'");
            System.out.println("Longitud: " + request.length() + " caracteres");
            
            String[] parts = request.split("\\|", -1);
            System.out.println("Partes encontradas: " + parts.length);
            for (int i = 0; i < parts.length; i++) {
                System.out.println("  Parte[" + i + "]: '" + parts[i] + "'");
            }
            
            if (parts.length == 0 || parts[0].isEmpty()) {
                out.println("ERROR|INVALID_REQUEST|Sin comando");
                return;
            }
            
            String command = parts[0].toUpperCase();
            
            switch (command) {
                case "TRAIN":
                    handleTraining(parts, out);
                    break;
                case "PREDICT":
                    handlePrediction(parts, out);
                    break;
                case "STATUS":
                    out.println("OK|" + raft.getState() + "|Leader:" + raft.getLeaderId());
                    break;
                case "REPLICATE":
                    if (parts.length > 1) {
                        handleReplication(parts[1]);
                    }
                    out.println("OK");
                    break;
                default:
                    out.println("ERROR|UNKNOWN_COMMAND|" + command);
            }
            
        } catch (SocketTimeoutException e) {
            System.err.println("Timeout leyendo del cliente");
            try {
                PrintWriter out = new PrintWriter(client.getOutputStream(), true);
                out.println("ERROR|TIMEOUT");
            } catch (IOException ioe) {
                // Ignorar
            }
        } catch (Exception e) {
            System.err.println("Error procesando cliente: " + e.getMessage());
            e.printStackTrace();
            try {
                PrintWriter out = new PrintWriter(client.getOutputStream(), true);
                out.println("ERROR|SERVER_ERROR|" + e.getMessage());
            } catch (IOException ioe) {
                // Ignorar
            }
        } finally {
            try {
                if (!client.isClosed()) {
                    client.close();
                }
            } catch (IOException e) {
                // Ignorar
            }
        }
    }
    
    private void handleTraining(String[] parts, PrintWriter out) {
        System.out.println("\n--- Procesando TRAIN ---");
        
        if (parts.length < 3) {
            String error = "Se requieren 3 partes: TRAIN|input_data|output_data. Recibidas: " + parts.length;
            System.err.println(error);
            out.println("ERROR|INVALID_REQUEST|" + error);
            return;
        }
        
        String inputData = parts[1];
        String outputData = parts[2];
        
        if (inputData.isEmpty() || outputData.isEmpty()) {
            out.println("ERROR|INVALID_REQUEST|Datos de entrada o salida vacíos");
            return;
        }
        
        String modelId = generateModelId();
        System.out.println("Generado model ID: " + modelId);
        System.out.println("Input data: " + inputData);
        System.out.println("Output data: " + outputData);
        
        try {
            MLModel model = new MLModel(modelId);
            model.train(inputData, outputData);
            storage.saveModel(modelId, model);
            System.out.println("Modelo entrenado y guardado exitosamente");
            
            // Replicar si es necesario
            String modelData = "MODEL|" + modelId + "|" + model.serialize();
            if (raft.isLeader()) {
                raft.replicate(modelData);
            }
            
            out.println("OK|" + modelId);
            System.out.println("Respuesta enviada: OK|" + modelId);
            
        } catch (Exception e) {
            String error = "Error entrenando: " + e.getMessage();
            System.err.println(error);
            e.printStackTrace();
            out.println("ERROR|TRAINING_FAILED|" + error);
        }
    }
    
    private void handlePrediction(String[] parts, PrintWriter out) {
        System.out.println("\n--- Procesando PREDICT ---");
        
        if (parts.length < 3) {
            String error = "Se requieren 3 partes: PREDICT|model_id|input_data. Recibidas: " + parts.length;
            System.err.println(error);
            out.println("ERROR|INVALID_REQUEST|" + error);
            return;
        }
        
        String modelId = parts[1];
        String inputData = parts[2];
        
        if (modelId.isEmpty()) {
            out.println("ERROR|INVALID_REQUEST|model_id vacío");
            return;
        }
        
        if (inputData.isEmpty()) {
            out.println("ERROR|INVALID_REQUEST|input_data vacío");
            return;
        }
        
        System.out.println("Model ID: " + modelId);
        System.out.println("Input data: " + inputData);
        
        MLModel model = storage.loadModel(modelId);
        if (model == null) {
            System.err.println("Modelo no encontrado: " + modelId);
            out.println("ERROR|MODEL_NOT_FOUND|" + modelId);
            return;
        }
        
        try {
            String prediction = model.predict(inputData);
            System.out.println("Predicción: " + prediction);
            out.println("OK|" + prediction);
            
        } catch (Exception e) {
            String error = "Error en predicción: " + e.getMessage();
            System.err.println(error);
            e.printStackTrace();
            out.println("ERROR|PREDICTION_FAILED|" + error);
        }
    }
    
    private void handleReplication(String data) {
        try {
            if (data != null && data.contains("|")) {
                String[] parts = data.split("\\|");
                if (parts[0].equals("MODEL") && parts.length >= 3) {
                    String modelId = parts[1];
                    String modelData = parts[2];
                    
                    MLModel model = new MLModel(modelId);
                    model.deserialize(modelData);
                    storage.saveModel(modelId, model);
                    System.out.println("Modelo " + modelId + " replicado exitosamente");
                }
            }
        } catch (Exception e) {
            System.err.println("Error en replicación: " + e.getMessage());
        }
    }
    
    private String generateModelId() {
        return "model_" + System.currentTimeMillis() + "_" + workerId;
    }
    
    public void shutdown() {
        running = false;
        try {
            if (serverSocket != null && !serverSocket.isClosed()) {
                serverSocket.close();
            }
        } catch (IOException e) {
            // Ignorar
        }
        executor.shutdown();
    }
    
    public static void main(String[] args) {
        if (args.length < 3) {
            System.err.println("Uso: Worker <workerId> <port> <monitorPort>");
            System.exit(1);
        }
        
        String workerId = args[0];
        int port = Integer.parseInt(args[1]);
        int monitorPort = Integer.parseInt(args[2]);
        
        System.out.println("=== Iniciando Worker ===");
        System.out.println("ID: " + workerId);
        System.out.println("Puerto: " + port);
        System.out.println("Puerto Monitor: " + monitorPort);
        
        Worker worker = new Worker(workerId, port, monitorPort);
        
        // Shutdown hook
        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            System.out.println("\nApagando worker " + workerId + "...");
            worker.shutdown();
        }));
        
        try {
            worker.start();
        } catch (Exception e) {
            System.err.println("Error fatal: " + e.getMessage());
            e.printStackTrace();
            System.exit(1);
        }
    }
}

File: python-worker/worker.py
import socket
import threading
import time
import os
import sys
from concurrent.futures import ThreadPoolExecutor

# Importar los otros módulos
try:
    from raft_consensus import RaftConsensus
    from http_monitor import HttpMonitor
    from ml_model import MLModel
    from storage import Storage
except ImportError as e:
    print(f"Error importando módulos: {e}")
    print("Asegúrate de que todos los archivos estén en el directorio")
    sys.exit(1)

class Worker:
    def __init__(self, worker_id, port, monitor_port):
        self.worker_id = worker_id
        self.port = port
        self.monitor_port = monitor_port
        self.storage = Storage(worker_id)
        self.raft = RaftConsensus(worker_id, port)
        self.monitor = HttpMonitor(monitor_port, self)
        self.executor = ThreadPoolExecutor(max_workers=10)
        self.running = True
        self.server_socket = None
    
    def configure_peers(self):
        """Configura los peers del RAFT según el worker_id"""
        if self.worker_id == "python1":
            self.raft.add_peer("java1", "java-worker-1:5000")
            self.raft.add_peer("java2", "java-worker-2:5001")
            self.raft.add_peer("python2", "python-worker-2:5003")
        elif self.worker_id == "python2":
            self.raft.add_peer("java1", "java-worker-1:5000")
            self.raft.add_peer("java2", "java-worker-2:5001")
            self.raft.add_peer("python1", "python-worker-1:5002")

    def start(self):
        print(f"Iniciando Worker {self.worker_id} en puerto {self.port}")
        
        # Configurar peers
        self.configure_peers()
        
        # Iniciar monitor HTTP en thread separado
        monitor_thread = threading.Thread(target=self.start_monitor, daemon=True)
        monitor_thread.start()
        
        # Iniciar RAFT
        raft_thread = threading.Thread(target=self.raft.start, daemon=True)
        raft_thread.start()
        
        # Esperar un poco
        time.sleep(2)
        
        # Iniciar servidor principal con reintentos
        max_retries = 5
        retry_count = 0
        
        while retry_count < max_retries:
            try:
                self.server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
                self.server_socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
                self.server_socket.bind(('0.0.0.0', self.port))
                self.server_socket.listen(5)
                print(f"Worker {self.worker_id} escuchando exitosamente en puerto {self.port}")
                break
            except OSError as e:
                if e.errno == 98:  # Address already in use
                    retry_count += 1
                    print(f"Puerto {self.port} en uso, reintento {retry_count}/{max_retries} en 5 segundos...")
                    time.sleep(5)
                else:
                    raise
        
        if retry_count >= max_retries:
            print(f"No se pudo iniciar el worker después de {max_retries} intentos")
            return
        
        # Loop principal
        while self.running:
            try:
                client, addr = self.server_socket.accept()
                self.executor.submit(self.handle_client, client)
            except Exception as e:
                if self.running:
                    print(f"Error aceptando conexión: {e}")
    
    def start_monitor(self):
        """Inicia el monitor HTTP con manejo de errores"""
        max_retries = 5
        retry_count = 0
        
        while retry_count < max_retries:
            try:
                self.monitor.start()
                break
            except OSError as e:
                if e.errno == 98:  # Address already in use
                    retry_count += 1
                    print(f"Puerto monitor {self.monitor_port} en uso, reintento {retry_count}/{max_retries}...")
                    time.sleep(5)
                else:
                    print(f"Error iniciando monitor: {e}")
                    break
    
    def handle_client(self, client):
        """Maneja las peticiones del cliente"""
        response = None
        try:
            # Recibir datos
            data = client.recv(4096).decode()
            if not data:
                client.close()
                return
            
            # Quitar saltos de línea y espacios
            data = data.strip()
            
            print(f"Worker {self.worker_id} recibió: {data}")
            parts = data.split('|')
            
            if len(parts) == 0:
                response = "ERROR|EMPTY_REQUEST"
            else:
                command = parts[0]
                
                # Procesar comando
                if command == 'TRAIN':
                    response = self.handle_training(parts)
                    print(f"DEBUG: handle_training retornó: {response}")
                elif command == 'PREDICT':
                    response = self.handle_prediction(parts)
                elif command == 'STATUS':
                    response = f"OK|{self.raft.get_state()}|Leader:{self.raft.leader_id}"
                elif command == 'REPLICATE':
                    if len(parts) > 1:
                        self.handle_replication(parts[1])
                    response = "OK"
                else:
                    response = f"ERROR|UNKNOWN_COMMAND|{command}"
            
            # CRÍTICO: Siempre enviar respuesta
            if response is None:
                response = "ERROR|NO_RESPONSE_GENERATED"
            
            print(f"Worker {self.worker_id} enviando respuesta: {response}")
            response_bytes = response.encode()
            total_sent = 0
            
            # Asegurar que se envíen todos los bytes
            while total_sent < len(response_bytes):
                sent = client.send(response_bytes[total_sent:])
                if sent == 0:
                    raise RuntimeError("Socket connection broken")
                total_sent += sent
            
            print(f"Worker {self.worker_id} respuesta enviada completamente ({total_sent} bytes)")
            
        except Exception as e:
            print(f"ERROR en handle_client: {e}")
            import traceback
            traceback.print_exc()
            
            # Intentar enviar error
            if response is None:
                try:
                    error_msg = f"ERROR|EXCEPTION|{str(e)}"
                    client.send(error_msg.encode())
                except:
                    print("No se pudo enviar mensaje de error")
        finally:
            print(f"Worker {self.worker_id} cerrando conexión con cliente")
            try:
                client.close()
            except:
                pass
    
    def handle_training(self, parts):
        """Maneja el entrenamiento de modelos"""
        print(f"DEBUG handle_training: Iniciando con parts={parts}")
        
        if len(parts) < 3:
            return "ERROR|INVALID_REQUEST|Se requieren 3 partes"
            
        model_id = self.generate_model_id()
        input_data = parts[1]
        output_data = parts[2]
        
        print(f"DEBUG: model_id={model_id}")
        
        try:
            # Entrenar el modelo localmente
            print("DEBUG: Creando MLModel")
            model = MLModel(model_id)
            
            print("DEBUG: Iniciando entrenamiento")
            model.train(input_data, output_data)
            
            print("DEBUG: Guardando modelo")
            self.storage.save_model(model_id, model)
            
            print(f"Modelo {model_id} entrenado exitosamente en {self.worker_id}")
            
            # Si somos líder, replicar (pero no bloquear)
            if self.raft.is_leader():
                try:
                    model_data = f"MODEL|{model_id}|{model.serialize()}"
                    self.raft.replicate(model_data)
                except Exception as e:
                    print(f"Error replicando: {e}")
            
            # IMPORTANTE: Siempre retornar respuesta
            result = f"OK|{model_id}"
            print(f"DEBUG handle_training: Retornando: {result}")
            return result
            
        except Exception as e:
            error_msg = f"Error en entrenamiento: {e}"
            print(error_msg)
            import traceback
            traceback.print_exc()
            return f"ERROR|TRAINING_FAILED|{str(e)}"
    
    def handle_prediction(self, parts):
        """Maneja las predicciones"""
        if len(parts) < 3:
            return "ERROR|INVALID_REQUEST|Se requieren 3 partes"
            
        model_id = parts[1]
        input_data = parts[2]
        
        model = self.storage.load_model(model_id)
        if not model:
            return f"ERROR|MODEL_NOT_FOUND|{model_id}"
        
        try:
            prediction = model.predict(input_data)
            return f"OK|{prediction}"
        except Exception as e:
            return f"ERROR|PREDICTION_FAILED|{str(e)}"
    
    def handle_replication(self, data):
        """Maneja la replicación de modelos"""
        try:
            if data and '|' in data:
                parts = data.split('|')
                if parts[0] == 'MODEL' and len(parts) >= 3:
                    model_id = parts[1]
                    model_data = parts[2]
                    model = MLModel(model_id)
                    model.deserialize(model_data)
                    self.storage.save_model(model_id, model)
                    print(f"Modelo {model_id} replicado en {self.worker_id}")
        except Exception as e:
            print(f"Error en replicación: {e}")
    
    def generate_model_id(self):
        """Genera un ID único para el modelo"""
        timestamp = int(time.time() * 1000)
        return f"model_{timestamp}_{self.worker_id}"
    
    def stop(self):
        """Detiene el worker"""
        self.running = False
        if self.server_socket:
            try:
                self.server_socket.close()
            except:
                pass

if __name__ == "__main__":
    if len(sys.argv) < 4:
        print("Uso: worker.py <worker_id> <port> <monitor_port>")
        sys.exit(1)
        
    worker_id = sys.argv[1]
    port = int(sys.argv[2])
    monitor_port = int(sys.argv[3])
    
    print(f"=== Iniciando Python Worker ===")
    print(f"ID: {worker_id}")
    print(f"Puerto principal: {port}")
    print(f"Puerto monitor: {monitor_port}")
    
    try:
        worker = Worker(worker_id, port, monitor_port)
        worker.start()
    except KeyboardInterrupt:
        print("\nDeteniendo worker...")
        worker.stop()
    except Exception as e:
        print(f"Error fatal: {e}")
        import traceback
        traceback.print_exc()
        sys.exit(1)